stan/math/prim/fun/to_int.hpp

```diff
@@ -0,0 +1,86 @@
+#ifndef STAN_MATH_PRIM_FUN_TO_INT_HPP
+#define STAN_MATH_PRIM_FUN_TO_INT_HPP
+
+#include <stan/math/prim/err/check_not_nan.hpp>
+#include <stan/math/prim/err/check_finite.hpp>
+#include <stan/math/prim/functor/apply_scalar_unary.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns the input scalar as an integer type. Specialisation for integral
+ * types which do not need conversion, reduces to a no-op.
+ *
+ * @tparam T type of integral argument
+ * @param x argument
+ * @return Input argument unchanged
+ */
+template <typename T, require_integral_t<T>* = nullptr>
+inline T to_int(T x) {
+  return std::forward<T>(x);
+}
+
+/**
+ * Returns the input scalar as an integer type. This function performs no
+ * rounding and simply truncates the decimal to return only the signficand as an
+ * integer.
+ *
+ * Casting NaN and Inf values to integers is considered undefined behavior as
+ * NaN and Inf cannot be represented as an integer and most implementations
+ * simply overflow, as such this function throws for these inputs.
+ *
+ * The function also throws for floating-point values that are too large to be
+ * represented as an integer.
+ *
+ * @tparam T type of argument (must be arithmetic)
+ * @param x argument
+ * @return Integer value of argument
+ */
+template <typename T, require_floating_point_t<T>* = nullptr>
+inline int to_int(T x) {
+  static const char* function = "to_int";
+  check_not_nan(function, "x", x);
+  check_finite(function, "x", x);
+  if (x < std::numeric_limits<int>::min()
```

Comment:
the check-finite test is redundant and the check-not-nan is also redundant if you rewrite as:

```cpp
if (!(x >= std::numeric_limits<int>::min() && x <= std::numeric_limits<int>::max())) { ... }
```
And the message should be that the value is not in range to be converted to integer (it's either too large if it's above max or too small if it's below min and it doesn't fit if it's NaN).